### ISSUE TITLE
Rewrite ListCtrl

### ISSUE THREAD

**Issue by Utumno** [wrye-bash/wrye-bash#255]
ModList.PopulateItems with:

```
         154268 function calls in 1.602 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    1.602    1.602 <string>:1(<module>)
        1    0.001    0.001    1.602    1.602 balt.py:1952(_pop)
        1    0.000    0.000    1.302    1.302 balt.py:2081(autosizeColumns)
       10    0.000    0.000    1.302    0.130 _controls.py:4550(SetColumnWidth)
       10    1.302    0.130    1.302    0.130 {wx._controls_.ListCtrl_SetColumnWidth}
      263    0.012    0.000    0.292    0.001 balt.py:1925(PopulateItem)
      263    0.001    0.000    0.070    0.000 __init__.py:798(<lambda>)
      263    0.008    0.000    0.070    0.000 __init__.py:828(setUI)
      263    0.002    0.000    0.069    0.000 __init__.py:3285(cachedCrc)
      526    0.002    0.000    0.052    0.000 bolt.py:737(getmtime)
      263    0.001    0.000    0.046    0.000 __init__.py:793(<lambda>)
      789    0.001    0.000    0.039    0.000 bolt.py:879(isdir)
      789    0.038    0.000    0.038    0.000 {nt._isdir}
      789    0.035    0.000    0.035    0.000 {nt.stat}
      263    0.001    0.000    0.035    0.000 balt.py:1747(InsertListCtrlItem)
     2367    0.002    0.000    0.027    0.000 _controls.py:4576(SetStringItem)
      263    0.001    0.000    0.027    0.000 bolt.py:707(size)
     2367    0.025    0.000    0.025    0.000 {wx._controls_.ListCtrl_SetStringItem}
      526    0.001    0.000    0.025    0.000 genericpath.py:60(getmtime)
      526    0.000    0.000    0.023    0.000 _controls.py:4572(SetItem)
      526    0.023    0.000    0.023    0.000 {wx._controls_.ListCtrl_SetItem}
      526    0.001    0.000    0.021    0.000 __init__.py:3431(getPath)
      526    0.001    0.000    0.018    0.000 __init__.py:3081(getPath)
      526    0.002    0.000    0.017    0.000 bolt.py:832(join)
      263    0.000    0.000    0.015    0.000 _controls.py:4759(InsertStringItem)
      263    0.015    0.000    0.015    0.000 {wx._controls_.ListCtrl_InsertStringItem}
     1841    0.006    0.000    0.013    0.000 bolt.py:1767(getItem)
      263    0.001    0.000    0.013    0.000 __init__.py:795(<lambda>)
      263    0.001    0.000    0.012    0.000 genericpath.py:55(getsize)
      885    0.002    0.000    0.012    0.000 bolt.py:473(GPath)
      263    0.001    0.000    0.012    0.000 balt.py:2331(_round)
      263    0.001    0.000    0.011    0.000 bolt.py:152(formatInteger)
      263    0.002    0.000    0.010    0.000 __init__.py:3162(getStatus)
      526    0.001    0.000    0.010    0.000 _controls.py:4566(GetItem)
      507    0.002    0.000    0.009    0.000 {method 'sort' of 'list' objects}
      718    0.005    0.000    0.009    0.000 ntpath.py:415(normpath)
     8591    0.006    0.000    0.009    0.000 bolt.py:1040(__cmp__)
      263    0.000    0.000    0.009    0.000 _controls.py:4588(SetItemImage)
      263    0.008    0.000    0.008    0.000 {wx._controls_.ListCtrl_SetItemImage}
      526    0.008    0.000    0.008    0.000 {wx._controls_.ListCtrl_GetItem}
        1    0.000    0.000    0.007    0.007 balt.py:2152(SortItems)
....
```

 and without autosizeColumns:

```
         154216 function calls in 0.267 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.267    0.267 <string>:1(<module>)
        1    0.001    0.001    0.267    0.267 balt.py:1952(_pop)
      263    0.011    0.000    0.257    0.001 balt.py:1925(PopulateItem)
      263    0.001    0.000    0.072    0.000 __init__.py:798(<lambda>)
      263    0.002    0.000    0.070    0.000 __init__.py:3285(cachedCrc)
      263    0.008    0.000    0.058    0.000 __init__.py:828(setUI)
      526    0.002    0.000    0.054    0.000 bolt.py:737(getmtime)
      263    0.001    0.000    0.047    0.000 __init__.py:793(<lambda>)
      789    0.001    0.000    0.040    0.000 bolt.py:879(isdir)
      789    0.039    0.000    0.039    0.000 {nt._isdir}
      789    0.036    0.000    0.036    0.000 {nt.stat}
      263    0.001    0.000    0.027    0.000 bolt.py:707(size)
      526    0.001    0.000    0.025    0.000 genericpath.py:60(getmtime)
      263    0.001    0.000    0.022    0.000 balt.py:1747(InsertListCtrlItem)
      526    0.001    0.000    0.021    0.000 __init__.py:3431(getPath)
      526    0.001    0.000    0.018    0.000 __init__.py:3081(getPath)
      526    0.002    0.000    0.017    0.000 bolt.py:832(join)
     2367    0.002    0.000    0.016    0.000 _controls.py:4576(SetStringItem)
     2367    0.014    0.000    0.014    0.000 {wx._controls_.ListCtrl_SetStringItem}
     1841    0.006    0.000    0.014    0.000 bolt.py:1767(getItem)
      263    0.001    0.000    0.013    0.000 __init__.py:795(<lambda>)
      263    0.001    0.000    0.013    0.000 genericpath.py:55(getsize)
      885    0.002    0.000    0.012    0.000 bolt.py:473(GPath)
      263    0.001    0.000    0.012    0.000 balt.py:2331(_round)
      263    0.000    0.000    0.011    0.000 _controls.py:4759(InsertStringItem)
      263    0.011    0.000    0.011    0.000 {wx._controls_.ListCtrl_InsertStringItem}
      507    0.004    0.000    0.011    0.000 {method 'sort' of 'list' objects}
      263    0.001    0.000    0.011    0.000 bolt.py:152(formatInteger)
      263    0.002    0.000    0.011    0.000 __init__.py:3162(getStatus)
      526    0.000    0.000    0.010    0.000 _controls.py:4572(SetItem)
      526    0.010    0.000    0.010    0.000 {wx._controls_.ListCtrl_SetItem}
      718    0.005    0.000    0.009    0.000 ntpath.py:415(normpath)
        1    0.000    0.000    0.009    0.009 balt.py:2152(SortItems)
......
```

That's the last thing I needed to see to be convinced that the list control we use (and the "ListCtrlAutoWidthMixin") needs replacing. Hell, it does not even have a GetSelected API call - let alone SortItems

However the code is **NOT** yet ready for substituting a new ListControl in - ObjectView being a candidate (in a nice submodule)

If anyone wants to start investigating be my guest - _do not spam this issue however with opinions and chit chat or have your post deleted_. I want this here to hash up a technical specification.

Meanwhile: @wrye-bash/active, @wrye-bash/bashers avoid setting the columns width to auto....

> [wxListCtrl has a really bad API design, but it comes from Microsoft so we can't complain too loudly about it](http://wxpython-users.1045709.n5.nabble.com/how-to-find-onto-all-the-items-of-a-ListCtrl-td2355029.html)

Some refactoring for current control
- [ ] **balt.ListCtrl**: 
  - move low level GUI stuff there and turn UIList to the data aware part of the class hierarchy (complete with UI Cache) ?
  - drop DnD callbacks - the fnXXX ones - just override OnDropXXX (?)
  - move style params there?

**Comment by Utumno** [comment]
Moving to the backburner for now

**Comment by Arthmoor** [comment]
Just curious, but if you're moving something to the "back burner" why would that also mean closing the ticket? Won't it just end up forgotten that way?

**Comment by Utumno** [comment]
No - I want it out of my face - backburner is kind of limbo between closed and open.
The tracker here could be better plus I am buried alive under a growing number of issues - the project _needs developers_. Meanwhile I am cleaning up, I just can't browse anymore in this chaos - with 307 I hope to close some long standing refactoring issues - meaning remove the backburner label cause I send them there. Taking care of the tracker sucks up a lot of my energy.

In short backburner may well mean _feel free to take those up_. Will have to update the readme on contributing.

**Comment by Utumno** [comment]
Seems I felt really alone at the time lol!

Anyway - apparently autosize columns took its share in boot also @Infernio - probably  #529 (solid 309 territory)

**Comment by Infernio** [comment]
Yup, marked as duplicate of #529, which will almost certainly be a rewrite :P
